command_line_tools.md

```diff
@@ -0,0 +1,71 @@
+# Command Line Tools
+
+When writing command line tools, there are a couple of pitfalls to be aware of.
+In the past we made certain painful experiences which we would like to prevent
+in the future.
+
+### Project Structure
+
+Let's consider an example project for our new command line tool called
+`cooltool`. The project structure should look something like this.
+
+```
+$ tree cooltool -d
+cooltool
+└── cmd
+    ├── create
+    │   └── cluster
+    │       ├── defaulting
+    │       └── validation
+    └── version
+
+6 directories
+```
+
+The structure tells us what kind of commands the tool provides.
+
+```
+cooltool version
+cooltool create cluster
+```
+
+Note the folders `defaulting` and `validation`. There isolated code for
+defaulting and validation should be located at. Having such code separated makes
+isolated tests for specific aspects of the required business logic more obvious.
+Isolation, dependency injection and other patterns should guide us here, even in
+command line tools.
+
+Each command folder should at least contain the following files.
+
+* `command.go`, the Cobra command setup.
+* `error.go`, error types and their matchers.
+* `flag.go`, flag initialization and validation.
+* `runner.go`, the command execution specific code, without the Cobra command setup.
```

Comment:
Should we always have a runner? The `create` command might not have any functionality other than the cobra structure here.

Reply:
> Should we always have a runner?

Yes. 

> The `create` command might not have any functionality other than the cobra structure here.

The point of the runner as I understood is to decouple the Cobra magic with the rest, whatever the rest is. It would be easiest to simply have it consistent. I would personally still go with the runner even if the command only prints a message. Further, these types of commands should be the minority, since most commands do real work with real business logic.

Reply:
I think having runner.go always is a good step towards having eventual code generation.